gadget/device.go

```diff
@@ -73,3 +75,25 @@ func FindDeviceForStructure(ps *PositionedStructure) (string, error) {
 
 	return found, nil
 }
+
+// encodeLabel encodes a name for use a partition or filesystem label symlink by
+// udev. The result matches the output of blkid_encode_string().
+func encodeLabel(in string) string {
+	const allowed = `#+-.:=@_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789`
+
+	buf := &bytes.Buffer{}
+
+	for _, r := range in {
+		switch {
+		case utf8.RuneLen(r) > 1:
```

Comment:
What happens if the label or partlabel uses multibyte characters with an encoding that's not utf-8 (or any invalid utf-8)?

Reply:
I think we do not have to care for other encodings. We're passing a string as input, which already is a valid utf-8 sequence. The source is a gadget YAML, which again, uses utf-8. If I understand the plumbing in systemd and udev, but the time any links are created by udev, everything is utf-8 only.

Reply:
A scenario that would potentially give us some problem is when you set your ext2/3/4 filesystem label to something containing utf-8 multibyte characters with a total byte size that's larger than the available space in such a way that it would be truncated to  16 bytes in the middle of a character code. When recovered, it could potentially lead to invalid utf-8. I didn't actually try this situation to see if something wrong happens, but as you mentioned before it would happen at recovery time when building the input string and woudn't even reach `encodeLabel`.